### ISSUE TITLE
sound error with pyglet

### ISSUE THREAD

**Issue by ghost** [pythonarcade/arcade#884]
This might have to dow w/ permissions on my computer, but just in case someone knows something - 
OSError: [WinError -2003435519] Windows Error 0x88960001

I am getting this error whenever i play sound using arcade library.
there is no error in my code - i made sure only like 28 times

**Comment by Cleptomania** [comment]
Can you provide some more info on this?

What version of Windows are you on?
What hardware are you on?
What type of file are you trying to play?

This is an invalid API call error for XAudio2, which is hard to avoid and generally caused by something specific in someone's setup.

**Comment by ghost** [comment]
@Cleptomania 
windows: Microsoft Windows [Version 10.0.17763.1217]
hardware: core i5
files: i have tried playing .mp3, .wav, .ogg files, none of them work

and some other times, nothing happens :(

**Comment by Aganthor** [comment]
I can confirm this bug too. Here are my specs:
Windows version: Windows 10 Enterprise LTSC 64 bit
CPU: Intel i7-8565U
Sound card: Realtek with driver version 10.0.17763.1

Python version 3.8.3
Pyglet version 1.5.16
arcade version 2.5.6

Here is the output from python:
```
Traceback (most recent call last):
  File "C:/Users/bergeronl/AppData/Roaming/JetBrains/PyCharmCE2020.1/scratches/correction_tp.py", line 187, in <module>
    main()
  File "C:/Users/bergeronl/AppData/Roaming/JetBrains/PyCharmCE2020.1/scratches/correction_tp.py", line 184, in main
    arcade.run()
  File "C:\Users\bergeronl\Documents\programmation\python\sprite_tutorial\venv\lib\site-packages\arcade\window_commands.py", line 274, in run
    pyglet.app.run()
  File "C:\Users\bergeronl\Documents\programmation\python\sprite_tutorial\venv\lib\site-packages\pyglet\app\__init__.py", line 107, in run
    event_loop.run()
  File "C:\Users\bergeronl\Documents\programmation\python\sprite_tutorial\venv\lib\site-packages\pyglet\app\base.py", line 167, in run
    timeout = self.idle()
  File "C:\Users\bergeronl\Documents\programmation\python\sprite_tutorial\venv\lib\site-packages\pyglet\app\base.py", line 237, in idle
    redraw_all = self.clock.call_scheduled_functions(dt)
  File "C:\Users\bergeronl\Documents\programmation\python\sprite_tutorial\venv\lib\site-packages\pyglet\clock.py", line 292, in call_scheduled_functions
    item.func(now - item.last_ts, *item.args, **item.kwargs)
  File "C:\Users\bergeronl\Documents\programmation\python\sprite_tutorial\venv\lib\site-packages\pyglet\media\drivers\xaudio2\interface.py", line 99, in _check_state
    self._create_xa2()
  File "C:\Users\bergeronl\Documents\programmation\python\sprite_tutorial\venv\lib\site-packages\pyglet\media\drivers\xaudio2\interface.py", line 118, in _create_xa2
    lib.XAudio2Create(ctypes.byref(self._xaudio2), 0, self.processor)
  File "_ctypes/callproc.c", line 948, in GetResult
OSError: [WinError -2003435519] Windows Error 0x88960001
```

**Comment by einarf** [comment]
This might be a problem in pyglet or something that is not supported? Got any sample files with this problem?

**Comment by ghost** [comment]
import arcade


sample_sound = arcade.load_sound(":resources:sounds/coin1.wav")
arcade.play_sound(sample_sound)


class SampleSound(arcade.Sound):
    def __init__(self):
        super().__init__(":resources:sounds/coin1.wav")


sound = SampleSound()
sound.play()


@einarf try this file - when i play it, the output is "process finished with exit code zero"
but theres no sound. sometimes this happens, and sometimes then windows error occurs
could you make sure its only a problem for some people? im pretty sure theres no error

**Comment by ghost** [comment]
and i get the same error as @Aganthor when i try to run the sound demo on
https://arcade.academy/examples/sound_demo.html#id2
i didnt alter ANY of the code, i just run it and in the end i get the

lib.XAudio2Create(ctypes.byref(self._xaudio2), 0, self.processor)
  File "_ctypes/callproc.c", line 935, in GetResult
OSError: [WinError -2003435519] Windows Error 0x88960001

error :( this has worked before idk why its not working now

**Comment by Aganthor** [comment]
> 
> 
> import arcade
> 
> sample_sound = arcade.load_sound(":resources:sounds/coin1.wav")
> arcade.play_sound(sample_sound)
> 
> class SampleSound(arcade.Sound):
> def **init**(self):
> super().**init**(":resources:sounds/coin1.wav")
> 
> sound = SampleSound()
> sound.play()
> 
> @einarf try this file - when i play it, the output is "process finished with exit code zero"
> but theres no sound. sometimes this happens, and sometimes then windows error occurs
> could you make sure its only a problem for some people? im pretty sure theres no error

I confirm the same results when I run this code sample. No sound is heard and exit code is zero.

**Comment by einarf** [comment]
Did something break in the last couple of releases? Try with pyglet 1.5.15 or upgrade to 1.5.17?

**Comment by ghost** [comment]
@einarf no... :( i think it might have to do with permissions on my computer, its fine

**Comment by einarf** [comment]
There are more people having this issue so we should keep this issue open. We need more information related to this issue since it seesm to be a case of sound working perfectly or not working at all.

**Comment by Cleptomania** [comment]
I’d like to see if someone who is experiencing this can test it out with the development branch, perhaps this is not an issue with Pyglet 2.

It seems to probably be a driver compatibility issue because I’ve been unable to reproduce it. 

Regardless in Arcade 2.6 I’m going to add a way to select which audio driver you want to prioritize so if people are having issues with XAudio2 they can force it to use DirectSound instead. Right now there’s no way to do that, and I think it would probably have to be done through environment variables or something as Arcade needs to set it before it imports Pyglet, which means end users would need to set it before they import Arcade.

**Comment by ghost** [comment]
upgrading and downgrading to 1.5.15 and 1.5.17 didnt work either :(
the weird thing is it used to work just fine

**Comment by Cleptomania** [comment]
@24wuw In 2.5 we completely re-did the sound module of Arcade. It maintained API consistency so for most users you wouldn't have noticed it changed, so if it worked pre 2.5 that would explain why it broke now.

Can you try installing arcade in a virtual environment from the development branch and running the sound demo using the below commands?

```
pip install -I https://github.com/pythonarcade/arcade/archive/refs/heads/development.zip
python -m arcade.examples.sound_demo
```

and report if you still have the same problems. I'm curious if this is possibly resolved in Pyglet 2

**Comment by ghost** [comment]
hmmmm... it does not work either... :(
when is pyglet 2 coming out, do you know?

when did 2.5 come out? i think it stopped working some time ago, maybe around halloween last year

**Comment by einarf** [comment]
Do you have a sample audio file that doesn't work? I think that is the first start here..

**Comment by ghost** [comment]
it is everything that doesnt work, including m4a, mp3, wav, and ogg. (these are the ones i've tried)

**Comment by einarf** [comment]
We're discussing this on the pyglet discord. It might be more related to the xaudio2 version available on the system.

**Comment by caffeinepills** [comment]
You are erroring on XAudio2 driver creation, which means your system doesn't support it. (Which is bizarre because Windows 10 ships with Xaudio2.) It's possible you have a corrupted version or maybe DirectX installer overwrote it, or LTSC has some version that differs from normal versions of Windows 10. In any case, if it's not supported you can bypass the audio driver to the older directsound at the top of your program with this method for the time being. 

```
import pyglet
pyglet.options['audio'] = ('directsound', 'xaudio2', 'openal', 'pulse', 'silent')
```

EDIT: I would also recommend trying `pyglet.options["debug_media"] = True` to see if you get any other errors that may shed light on it.

In the next pyglet release, we will add a try/except on it to fallback to directsound, to prevent hard crashes on a case like this.

As far as why your example doesn't play audio is because your program needs the loop to play audio. You need to have an event loop to keep the program from closing as well as to schedule the samples to be played (for directsound).  I believe in arcade the equivalent is `arcade.run()`

**Comment by einarf** [comment]
Looks like there is a possible fix incoming for Windows 10 Enterprise LTSC

**Comment by ghost** [comment]
hooray!!! yes it does work now
but the thing is, you can't do functions like is_complete(), get_volume(), and get_stream_position()
the parameter is supposed to be a 'player'.... and theres a default in the arcade.Sound class, but the override @caffeinepills gave apparently just messes it up 😢

**Comment by einarf** [comment]
> hooray!!! yes it does work now

Can you describe what you did to make this work?

**Comment by ghost** [comment]
> > hooray!!! yes it does work now
> 
> Can you describe what you did to make this work?

np, i just did what @caffeinepills said to do - right after you import arcade, import pyglet (useless import, as arcade calls pyglet, but it won't work if you dont) and then just do the line `pyglet.options['audio'] = ('directsound', 'xaudio2', 'openal', 'pulse', 'silent')`

however, like i said, some builtin functions dont work - but its a lot better than nothing

**Comment by Cleptomania** [comment]
I'm not sure exactly how everything would work out doing that or not.

It's my understanding the audio backend order needs set before importing `pyglet.media`.

In Arcade's sound module we set the backend order, and then import it, so I'm not 100% certain which would take precedence, and in what scenarios Arcade's would versus a user-defined one.

**Comment by Aganthor** [comment]
Hello!

I tried what @caffeinepills suggested. Here is my code sample : 
```
import arcade
import pyglet

pyglet.options['audio'] = ('directsound', 'xaudio2', 'openal', 'pulse', 'silent')
pyglet.options["debug_media"] = True

sample_sound = arcade.load_sound(":resources:sounds/coin1.wav")
arcade.play_sound(sample_sound)
```

Here is the output: 
```
Constructing DirectSoundDriver
refill, write_size = 88200
Getting new audio data buffer.
Dispatching pending events: []
Remaining events: []
New audio data available: 17296 bytes
write 17296
DirectSoundBuffer.lock(0, 17296)
refill, write_size = 70904
Getting new audio data buffer.
Dispatching pending events: []
Remaining events: []
No more audio data.
write silence
DirectSoundBuffer.lock(17296, 70904)
Dispatching pending events: []
Remaining events: []
DirectSound play
Getting new audio data buffer.
Dispatching pending events: []
Remaining events: []
return DirectSound play

Process finished with exit code 0
```

I just hear a small clicking sound and then nothing. No sound at all.

I also tried to run this:
```
import pyglet

pyglet.options['audio'] = ('directsound', 'xaudio2', 'openal', 'pulse', 'silent')
pyglet.options["debug_media"] = True

explosion = pyglet.media.load('09 Explosion - debris falling.wav', streaming=False)
explosion.play()
```

And I also have no sound.

**Comment by ghost** [comment]
@Aganthor dont forget the line arcade.run(). @caffeinepills reminded to do that
and dont do `pyglet.options["debug_media"] = True`, i didnt and it worked for me

**Comment by Aganthor** [comment]
> 
> 
> @Aganthor dont forget the line arcade.run(). @caffeinepills reminded to do that
> and dont do `pyglet.options["debug_media"] = True`, i didnt and it worked for me

Ah ha! That worked. I now hear the coin sound with the arcade sample code. I'll have to try and figure out the pyglet code snippets.

**Comment by ghost** [comment]
yes, it does work 😄 - however, manually overriding the pyglet causes some problems, and you can't do some built in arcade functions that you could before. but like i said, its better than nothing. hopefully in the next pyglet or arcade update this issue would be resolved

**Comment by Cleptomania** [comment]
This should be resolved in Arcade 2.6. Pyglet has implemented an automatic fallback to DirectSound if XAudio2 is unavailable.

There is also some talk of adding in some new codecs to Pyglet or to Arcade that we can ship with the library to provide as fallbacks for systems that might be missing codecs like Windows N versions.

**Comment by ghost** [comment]
seems like theres nothing we can do rn

**Comment by ghost** [comment]
correction - using pygame's mixer works pretty seamlessly with arcade, i wouldnt use the playsound module though as it pauses your game/animation

its pretty good for now, until this issue is resolved

**Comment by Cleptomania** [comment]
This is resolved in 2.6, and we have the 2.6.0.dev2 pre-release out now so I'm gonna go ahead and close this.
